lib/jsonapi/rails.rb

```diff
@@ -62,8 +62,22 @@ def self.add_errors_renderer!
             details[attr] ||= []
             details[attr] << error.detail.merge(message: error.message)
           end
-        elsif resource.respond_to?(:details)
-          details = resource.details
+        elsif resource.respond_to?(:details) && resource.respond_to?(:messages)
+          resource.details.each do |attr, problems|
+            problems.each_with_index do |error, index|
+              details[attr] ||= []
+
+              if error[:error].is_a?(Hash)
+                current = error[:error].dup
+                current[:error] ||= :invalid
+
+                details[attr] << current
+              else
+                message = resource.messages[attr][index]
+                details[attr] << error.merge(message: message)
+              end
+            end
+          end
```

Comment:
@johvet would you be kind to explain or provide a test for this changeset please. As far as I know, the error handling is supported for rails 4-6...

Reply:
I am terribly sorry for my late reply. I don't recall the very details, but we have noticed that the current solution in master does not work consistently between Rails 5.x and 6.x -- It also makes a difference on whether symbols are used or text messages.

Reply:
IIRC it's because the previous implementation didn't work for something like:

```
errors.add(:foo, :blank)
errors.add(:foo, :some_custom_key)
```

but did work with

```
errors.add(:foo, 'some string')
```

IIRC you'd end up with `message: 'some_custom_key'` instead of the I18n'ed string that exists in `resource.messages`. I think this is a Rails bug though.

But looking above, the `current[:error] ||= :invalid` is redundant/impossible, so I think this needs some comments explaining the different ways the data might be available